# Exercise 4.7: Deriving New Variables
#### by Claudia Lan Yee Chiu

### Importing Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

### Importing Dataframes

In [3]:
# Assign path
path = r'C:\Users\clany\OneDrive\Documents\CareerFoundry\Chap_4_Pythons Fundamentals for Data Analysts\Instacart Basket Analysis'

In [4]:
# Importing ords_prods_merged pkl
df_ords_prods_merged = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data','orders_products_merged.pkl'))

In [7]:
# Work with a subset of the dataframe ords_prods_merged: first million rows only
df = df_ords_prods_merged[:1000000]

In [9]:
# Check the size of subset
df.shape

(1000000, 15)

### If-Statements with User-Defined Functions

#### Create a ´Price_label´: 3 different prices ranges

In [10]:
# Defining a Price_label function

def price_label(row):
    
    if row['prices'] <= 5:
        return 'Low range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid range product'
    elif row['prices'] > 15:
        return 'High range product'
    else: return 'Not enough data'

In [11]:
# Apply the new function

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\clany\AppData\Local\Temp\ipykernel_31536\4061432024.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [12]:
df['price_range'].value_counts(dropna = False)

Mid range product    756450
Low range product    243550
Name: price_range, dtype: int64

In [13]:
# Check what is the most expensive product 
df['prices'].max()

14.8

### If-Statements with the loc() Function

#### Applying the loc() function instead of if-then-else

In [14]:
# Applying above function in single cells using loc()
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\clany\AppData\Local\Temp\ipykernel_31536\2787218310.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [15]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [16]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [17]:
df['price_range_loc'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range_loc, dtype: int64

#### Applying the loc() function on the entire dataframe ords_prods_merged

In [21]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, ' price_range_loc'] = 'High-range product'

##### I ran the below function which caused my laptop to crash, and will hence not try again for now
##### df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

### If-Statements with For-Loops

#### Create a new column in ords_prods_merged to summarize how busy each day of the week is.
##### Start to find out on which day mmost orders take place

In [23]:
# Find the frequency of the ´orders_day_of_the_week`column
df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

##### 0 = Saturday, ie highest frequency. 4 = Wednesday, ie lowest frequency day

In [25]:
# Create a new column with one of three different values ´Busiest / Least / Regularly  busy´
result = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [26]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [27]:
# As seen above, the result is not very helpful! To make sense, combine it by creating a new column with ´result´ in the dataframe
# Add new column ´busiest_day´ = result

df_ords_prods_merged['busiest_day'] = result

In [28]:
# Check value counts

df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

## Task 4.7 Deriving New Variables

### Step 2: New Column for two busiest and two slowest days of the week

In [29]:
# First, find the busiest and slowest days from the ´orders_day_of_the_week`column
df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

#### The two busiest days are Saturday (0) and Sunday (1)
#### The two slowest days are Wednesday (4) and Tuesday (3)

In [32]:
# Second: define the for-loop and create an empty list called ´result_days´

result_days = []

for value in df_ords_prods_merged['orders_day_of_week']:
    if value == 0 or value == 1:
        result_days.append('Busiest days')
    elif value == 4 or value == 3:
        result_days.append('Slowest days')
    else:
        result_days.append('Normal days')

In [33]:
# Add result_days to dataframe in a new column
df_ords_prods_merged['Busiest_days'] = result_days

#### Step 3: Check the values of this new column for accuracy

In [34]:
# Check the header to view the newly created column
df_ords_prods_merged.head()

order_id  user_id eval_set  number_of_orders  orders_day_of_week  \
0   2539329        1    prior                 1                   2   
1   2398795        1    prior                 2                   3   
2    473747        1    prior                 3                   3   
3   2254736        1    prior                 4                   4   
4    431534        1    prior                 5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

   price_range_loc  price_range_loc     busiest_day  Busiest_days  
0              NaN              NaN  Regularly busy   Normal days  
1              NaN              NaN  Regularly busy  Slowest days  
2              NaN              NaN  Regularly busy  Slowest days  
3              NaN              NaN      Least busy  Slowest days  
4              NaN              NaN      Least busy  Slowest days

In [35]:
# Check for accuracy
df_ords_prods_merged['Busiest_days'].value_counts(dropna = False)

Normal days     12916111
Busiest days    11864412
Slowest days     7624336
Name: Busiest_days, dtype: int64

###### Observation: when adding the results of group 0 & 1 from previous function, 6204182 + 5660230 it returns the same as above digit
###### under the ´busiest days´ ie, 11864412. Same for the Slowest days, which is the sum of group 3 & 4, ie, 7624336: hence the loop is correct

#### Step 4: Create a new column containing labels `Busiest_period_of_day´

In [38]:
# Check the required column on which to focus first
df_ords_prods_merged.head()

order_id  user_id eval_set  number_of_orders  orders_day_of_week  \
0   2539329        1    prior                 1                   2   
1   2398795        1    prior                 2                   3   
2    473747        1    prior                 3                   3   
3   2254736        1    prior                 4                   4   
4    431534        1    prior                 5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

   price_range_loc  price_range_loc     busiest_day  Busiest_days  
0              NaN              NaN  Regularly busy   Normal days  
1              NaN              NaN  Regularly busy  Slowest days  
2              NaN              NaN  Regularly busy  Slowest days  
3              NaN              NaN      Least busy  Slowest days  
4              NaN              NaN      Least busy  Slowest days

In [39]:
# Check the frequency of values in the "order_hour_of_day" column 
df_ords_prods_merged['order_hour_of_day'].value_counts()

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

#### The request is to have 3 periods of time, hence we need to divide the 24 entries (ie hours) into 3 labels of 8 values as such: “Most orders” (10,11,14,15,13,12,16,9), “Average orders” (17,8,18,19,20,7,21,22), and “Fewest orders” (23,6,0,1,5,2,4,3) 

In [40]:
# Create an If-statement with For-loop using the above 3 requested labels

time_of_orders = []

for value in df_ords_prods_merged['order_hour_of_day']:
    if value in [10, 11, 14, 15, 13, 12, 16, 9]:
        time_of_orders.append('Most Orders')
    elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
        time_of_orders.append('Fewest Orders')
    else:
        time_of_orders.append('Average Orders')

In [41]:
# Create a new column "Busiest_period_of_day" with the result printed in it

df_ords_prods_merged['busiest_period_of_day'] = time_of_orders

In [43]:
# Check above result
df_ords_prods_merged[['order_hour_of_day', 'busiest_period_of_day']].head(12)

order_hour_of_day busiest_period_of_day
0                   8        Average Orders
1                   7        Average Orders
2                  12           Most Orders
3                   7        Average Orders
4                  15           Most Orders
5                   7        Average Orders
6                   9           Most Orders
7                  14           Most Orders
8                  16           Most Orders
9                   8        Average Orders
10                  9           Most Orders
11                 16           Most Orders

#### Step 5: Print the frequency for this new column

In [45]:
# Check the frequency of the new column
df_ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

Most Orders       21118071
Average Orders     9997651
Fewest Orders      1289137
Name: busiest_period_of_day, dtype: int64

#### Step 7: Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder

In [47]:
# Export dataframe to the Prepared Data folder
df_ords_prods_merged.to_pickle(os.path.join(path, 'Data','Prepared Data', 'orders_products_merged_task4_7.pkl'))